platform/posix/transport/src/openssl_posix.c

```diff
@@ -0,0 +1,563 @@
+/* Standard includes. */
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+
+/* POSIX socket includes. */
+#include <limits.h>
+#include <errno.h>
+#include <netdb.h>
+#include <time.h>
+#include <poll.h>
+#include <unistd.h>
+#include <arpa/inet.h>
+
+#include <sys/socket.h>
+#include <sys/types.h>
+
+/* OpenSSL include. */
+#include <openssl/ssl.h>
+
+/* Transport interface include. */
+#include "transport_interface.h"
+
+#include "openssl_posix.h"
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Timeout for transport send.
+ *
+ * @note Setting to a negative value implies an infinite timeout.
+ */
+static int tlsSendTimeout = -1;
+
+/**
+ * @brief Timeout for transport recv.
+ *
+ * @note Setting to a negative value implies an infinite timeout.
+ */
+static int tlsRecvTimeout = -1;
+
+/**
+ * @brief Definition of the HTTP network context.
+ *
+ * @note For this transport implementation, the socket descriptor and
+ * SSL context is used.
+ */
+struct NetworkContext
+{
+    int tcpSocket;
+    SSL * pSslContext;
+};
+
+/**
+ * @brief A NULL-terminated string for the SNI host name.
+ *
+ * @note If SNI is enabled, this NULL-terminated string is set to
+ * #OpensslCredentials.sniHostName, then passed to #SSL_set_tlsext_host_name.
+ */
+static char sniHostName[ HOST_NAME_MAX + 1 ];
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Passes TLS credentials to the OpenSSL library.
+ *
+ * Provides the root CA certificate, client certificate, and private key to the
+ * OpenSSL library. If the client certificate or private key is not NULL, mutual
+ * authentication is used when performing the TLS handshake.
+ *
+ * @param[in] pSslContext Destination for the imported credentials.
+ * @param[in] pOpensslCredentials TLS credentials to be validated.
+ *
+ * @return 1 on success; otherwise, failure;
+ */
+static int readCredentials( SSL_CTX * pSslContext,
+                            OpensslCredentials_t * pOpensslCredentials );
+
+/*-----------------------------------------------------------*/
+
+static int readCredentials( SSL_CTX * pSslContext,
```

Comment:
A manual complexity count for this function gives 22, we are targeting <= 8.

Reply:
Although I do agree that complexity could be lower, we are following complexity / MISRA  criteria for libraries only.

Reply:
Complexity is now <= 8.